File: day10/main.go
package main

import (
	"bufio"
	"bytes"
	"fmt"
	"log"
	"os"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatalln(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatalln(err)
		}
	}()

	var grid [][]byte
	start := Coord{x: -1, y: -1}

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		k := []byte(scanner.Text())
		x := bytes.IndexByte(k, 'S')
		if x >= 0 {
			start = Coord{
				x: x,
				y: len(grid),
			}
		}
		grid = append(grid, k)
	}

	if err := scanner.Err(); err != nil {
		log.Fatalln(err)
	}

	gridBounds := Coord{
		x: len(grid[0]),
		y: len(grid),
	}
	simpleGrid := make([][]byte, gridBounds.y)
	for i := range simpleGrid {
		simpleGrid[i] = make([]byte, gridBounds.x)
	}
	bounds := make([]Coord, gridBounds.y)
	for i := range bounds {
		bounds[i] = Coord{
			x: gridBounds.x,
			y: -1,
		}
	}

	neighbors := getStartNeighbors(grid, start)
	if len(neighbors) != 2 {
		log.Fatalln("Invalid neighbors")
	}

	startChar, ok := getStartChar(neighbors[0].dir, neighbors[1].dir)
	if !ok {
		log.Fatalln("Invalid neighbors")
	}
	simpleGrid[start.y][start.x] = startChar
	bounds[start.y] = Coord{
		x: start.x,
		y: start.x,
	}
	for _, i := range neighbors {
		simpleGrid[i.coord.y][i.coord.x] = grid[i.coord.y][i.coord.x]
		b := bounds[i.coord.y]
		bounds[i.coord.y] = Coord{
			x: min(b.x, i.coord.x),
			y: max(b.y, i.coord.x),
		}
	}

	steps := 1
	for allCoordsNotSame(neighbors) {
		for i := range neighbors {
			next, ok := nextStep(grid, neighbors[i])
			if !ok {
				log.Fatalln("Invalid next step")
			}
			neighbors[i] = next
			simpleGrid[next.coord.y][next.coord.x] = grid[next.coord.y][next.coord.x]
			b := bounds[next.coord.y]
			bounds[next.coord.y] = Coord{
				x: min(b.x, next.coord.x),
				y: max(b.y, next.coord.x),
			}
		}
		steps++
	}
	fmt.Println("Part 1:", steps)

	sum := 0
	for n, i := range simpleGrid {
		b := bounds[n]
		if b.y < 0 {
			continue
		}
		sum += getInsideCount(i[b.x : b.y+1])
	}
	fmt.Println("Part 2:", sum)
}

func getInsideCount(row []byte) int {
	count := 0
	inside := false
	var prev byte = 0
	for _, i := range row {
		switch i {
		case 0:
			if inside {
				count++
			}
		case '|':
			inside = !inside
			prev = 0
		case '-':
		case 'L':
			prev = 'L'
		case 'J':
			switch prev {
			case 'L':
			case 'F':
				inside = !inside
			}
			prev = 0
		case '7':
			switch prev {
			case 'L':
				inside = !inside
			case 'F':
			}
			prev = 0
		case 'F':
			prev = 'F'
		}
	}
	return count
}

func allCoordsNotSame(neighbors []CoordDir) bool {
	coord := neighbors[0].coord
	for _, i := range neighbors {
		if i.coord != coord {
			return true
		}
	}
	return false
}

func nextStep(grid [][]byte, posdir CoordDir) (CoordDir, bool) {
	transform, ok := tileDirMap[grid[posdir.coord.y][posdir.coord.x]]
	if !ok {
		return CoordDir{}, false
	}
	nextDir := transform[posdir.dir]
	switch nextDir {
	case DirNorth:
		return CoordDir{
			coord: Coord{
				x: posdir.coord.x,
				y: posdir.coord.y - 1,
			},
			dir: nextDir,
		}, true
	case DirEast:
		return CoordDir{
			coord: Coord{
				x: posdir.coord.x + 1,
				y: posdir.coord.y,
			},
			dir: nextDir,
		}, true
	case DirSouth:
		return CoordDir{
			coord: Coord{
				x: posdir.coord.x,
				y: posdir.coord.y + 1,
			},
			dir: nextDir,
		}, true
	case DirWest:
		return CoordDir{
			coord: Coord{
				x: posdir.coord.x - 1,
				y: posdir.coord.y,
			},
			dir: nextDir,
		}, true
	default:
		return CoordDir{}, false
	}
}

type (
	Coord struct {
		x, y int
	}

	CoordDir struct {
		coord Coord
		dir   Dir
	}

	Dir int
)

const (
	DirNorth Dir = 0
	DirEast      = 1
	DirSouth     = 2
	DirWest      = 3
)

var tileDirMap = map[byte][4]Dir{
	'|': {DirNorth, -1, DirSouth, -1},
	'-': {-1, DirEast, -1, DirWest},
	'L': {-1, -1, DirEast, DirNorth},
	'J': {-1, DirNorth, DirWest, -1},
	'7': {DirWest, DirSouth, -1, -1},
	'F': {DirEast, -1, -1, DirSouth},
}

func getStartChar(a, b Dir) (byte, bool) {
	if a > b {
		a, b = b, a
	}
	switch a {
	case DirNorth:
		switch b {
		case DirEast:
			return 'L', true
		case DirSouth:
			return '|', true
		case DirWest:
			return 'J', true
		}
	case DirEast:
		switch b {
		case DirSouth:
			return 'F', true
		case DirWest:
			return '-', true
		}
	case DirSouth:
		switch b {
		case DirWest:
			return '7', true
		}
	case DirWest:
	}
	return 0, false
}

func getStartNeighbors(grid [][]byte, pos Coord) []CoordDir {
	h := len(grid)
	w := len(grid[0])
	neighbors := make([]CoordDir, 0, 4)
	if pos.x > 0 {
		// has left neighbor
		switch grid[pos.y][pos.x-1] {
		case '-', 'L', 'F':
			neighbors = append(neighbors, CoordDir{
				coord: Coord{
					x: pos.x - 1,
					y: pos.y,
				},
				dir: DirWest,
			})
		}
	}
	if pos.y > 0 {
		// has top neighbor
		switch grid[pos.y-1][pos.x] {
		case '|', '7', 'F':
			neighbors = append(neighbors, CoordDir{
				coord: Coord{
					x: pos.x,
					y: pos.y - 1,
				},
				dir: DirNorth,
			})
		}
	}
	if pos.x+1 < w {
		// has right neighbor
		switch grid[pos.y][pos.x+1] {
		case '-', 'J', '7':
			neighbors = append(neighbors, CoordDir{
				coord: Coord{
					x: pos.x + 1,
					y: pos.y,
				},
				dir: DirEast,
			})
		}
	}
	if pos.y+1 < h {
		// has bot neighbor
		switch grid[pos.y+1][pos.x] {
		case '|', 'L', 'J':
			neighbors = append(neighbors, CoordDir{
				coord: Coord{
					x: pos.x,
					y: pos.y + 1,
				},
				dir: DirSouth,
			})
		}
	}
	return neighbors
}

File: day10/src/main.rs
use once_cell::sync::Lazy;
use std::collections::HashMap;
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;
use std::ops::Index;
use std::ops::IndexMut;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let mut start = None;
    let grid = {
        let mut elements = Vec::new();
        let mut width = 0;
        for (y, line) in reader.lines().enumerate() {
            let k = line?;
            width = k.len();
            if let Some(x) = k.bytes().position(|v| v == b'S') {
                start = Some(Coord { x, y });
            }
            elements.extend(k.bytes());
        }
        Grid::new(elements, width)
    };
    let start = start.ok_or("Missing start symbol")?;
    let mut pipe_grid = Grid::new(vec![0; grid.width * grid.height], grid.width);
    let mut bounds = vec![None; grid.height];

    let (mut na, mut nb) = get_start_neighbors(&grid, start).ok_or("Invalid neighbors")?;
    {
        let start_char = get_start_char(na.dir, nb.dir).ok_or("Invalid neighbors")?;
        pipe_grid[start] = start_char;
        bounds[start.y] = Some(Coord {
            x: start.x,
            y: start.x,
        });
        pipe_grid[na.coord] = grid[na.coord];
        pipe_grid[nb.coord] = grid[nb.coord];
        match bounds[na.coord.y] {
            Some(Coord { x, y }) => {
                bounds[na.coord.y] = Some(Coord {
                    x: x.min(na.coord.x),
                    y: y.max(na.coord.x),
                });
            }
            None => {
                bounds[na.coord.y] = Some(Coord {
                    x: na.coord.x,
                    y: na.coord.x,
                });
            }
        }
        match bounds[nb.coord.y] {
            Some(Coord { x, y }) => {
                bounds[nb.coord.y] = Some(Coord {
                    x: x.min(nb.coord.x),
                    y: y.max(nb.coord.x),
                });
            }
            None => {
                bounds[nb.coord.y] = Some(Coord {
                    x: nb.coord.x,
                    y: nb.coord.x,
                });
            }
        }
    }
    let mut steps = 1;
    while na.coord != nb.coord {
        {
            let dir = TILE_DIR_MAP
                .get(&grid[na.coord])
                .ok_or("Invalid pipe path 1")?[na.dir]
                .ok_or("Invalid pipe path 2")?;
            let coord = na.coord.step(dir);
            na = CoordDir { coord, dir };
        }
        {
            let dir = TILE_DIR_MAP
                .get(&grid[nb.coord])
                .ok_or("Invalid pipe path 3")?[nb.dir]
                .ok_or("Invalid pipe path 4")?;
            let coord = nb.coord.step(dir);
            nb = CoordDir { coord, dir };
        }
        pipe_grid[na.coord] = grid[na.coord];
        pipe_grid[nb.coord] = grid[nb.coord];
        match bounds[na.coord.y] {
            Some(Coord { x, y }) => {
                bounds[na.coord.y] = Some(Coord {
                    x: x.min(na.coord.x),
                    y: y.max(na.coord.x),
                });
            }
            None => {
                bounds[na.coord.y] = Some(Coord {
                    x: na.coord.x,
                    y: na.coord.x,
                });
            }
        }
        match bounds[nb.coord.y] {
            Some(Coord { x, y }) => {
                bounds[nb.coord.y] = Some(Coord {
                    x: x.min(nb.coord.x),
                    y: y.max(nb.coord.x),
                });
            }
            None => {
                bounds[nb.coord.y] = Some(Coord {
                    x: nb.coord.x,
                    y: nb.coord.x,
                });
            }
        }
        steps += 1;
    }
    println!("Part 1: {}", steps);

    let pipe_grid = pipe_grid;

    let mut sum = 0;
    for i in 0..pipe_grid.height {
        if let Some(b) = bounds[i] {
            let a = pipe_grid.width * i;
            sum += get_inside_area(&pipe_grid.elements[a + b.x..a + b.y]);
        }
    }
    println!("Part 2: {}", sum);

    Ok(())
}

fn get_inside_area(row: &[u8]) -> usize {
    let mut count = 0;
    let mut inside = false;
    let mut prev = 0;
    for i in row {
        match i {
            0 => {
                if inside {
                    count += 1
                }
            }
            b'|' => {
                inside = !inside;
                prev = 0;
            }
            b'L' => prev = b'L',
            b'J' => {
                if prev == b'F' {
                    inside = !inside;
                }
                prev = 0;
            }
            b'7' => {
                if prev == b'L' {
                    inside = !inside;
                }
                prev = 0;
            }
            b'F' => prev = b'F',
            _ => {}
        }
    }
    count
}

struct TileDirMap {
    north: Option<Dir>,
    east: Option<Dir>,
    south: Option<Dir>,
    west: Option<Dir>,
}

impl Index<Dir> for TileDirMap {
    type Output = Option<Dir>;

    fn index(&self, dir: Dir) -> &Self::Output {
        match dir {
            Dir::North => &self.north,
            Dir::East => &self.east,
            Dir::South => &self.south,
            Dir::West => &self.west,
        }
    }
}

static TILE_DIR_MAP: Lazy<HashMap<u8, TileDirMap>> = Lazy::new(|| {
    HashMap::from([
        (
            b'|',
            TileDirMap {
                north: Some(Dir::North),
                east: None,
                south: Some(Dir::South),
                west: None,
            },
        ),
        (
            b'-',
            TileDirMap {
                north: None,
                east: Some(Dir::East),
                south: None,
                west: Some(Dir::West),
            },
        ),
        (
            b'L',
            TileDirMap {
                north: None,
                east: None,
                south: Some(Dir::East),
                west: Some(Dir::North),
            },
        ),
        (
            b'J',
            TileDirMap {
                north: None,
                east: Some(Dir::North),
                south: Some(Dir::West),
                west: None,
            },
        ),
        (
            b'7',
            TileDirMap {
                north: Some(Dir::West),
                east: Some(Dir::South),
                south: None,
                west: None,
            },
        ),
        (
            b'F',
            TileDirMap {
                north: Some(Dir::East),
                east: None,
                south: None,
                west: Some(Dir::South),
            },
        ),
    ])
});

#[derive(Debug, Clone, Copy)]
enum Dir {
    North,
    East,
    South,
    West,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
struct Coord {
    x: usize,
    y: usize,
}

impl Coord {
    fn step(&self, dir: Dir) -> Self {
        let &Self { x, y } = self;
        match dir {
            Dir::North => Self {
                x,
                y: y.wrapping_sub(1),
            },
            Dir::East => Self { x: self.x + 1, y },
            Dir::South => Self { x, y: self.y + 1 },
            Dir::West => Self {
                x: x.wrapping_sub(1),
                y,
            },
        }
    }
}

struct Grid {
    width: usize,
    height: usize,
    elements: Vec<u8>,
}

impl Grid {
    fn new(elements: Vec<u8>, width: usize) -> Self {
        Self {
            width,
            height: elements.len() / width,
            elements,
        }
    }
}

impl Index<Coord> for Grid {
    type Output = u8;

    fn index(&self, pos: Coord) -> &Self::Output {
        if pos.x >= self.width || pos.y >= self.height {
            &0
        } else {
            &self.elements[self.width * pos.y + pos.x]
        }
    }
}

impl IndexMut<Coord> for Grid {
    fn index_mut(&mut self, pos: Coord) -> &mut Self::Output {
        &mut self.elements[self.width * pos.y + pos.x]
    }
}

#[derive(Debug, Clone, Copy)]
struct CoordDir {
    coord: Coord,
    dir: Dir,
}

fn get_start_neighbors(grid: &Grid, pos: Coord) -> Option<(CoordDir, CoordDir)> {
    let mut neighbors = Vec::with_capacity(4);
    let top = pos.step(Dir::North);
    match grid[top] {
        b'|' | b'7' | b'F' => neighbors.push(CoordDir {
            coord: top,
            dir: Dir::North,
        }),
        _ => {}
    }
    let right = pos.step(Dir::East);
    match grid[right] {
        b'-' | b'J' | b'7' => neighbors.push(CoordDir {
            coord: right,
            dir: Dir::East,
        }),
        _ => {}
    }
    let bot = pos.step(Dir::South);
    match grid[bot] {
        b'|' | b'L' | b'J' => neighbors.push(CoordDir {
            coord: bot,
            dir: Dir::South,
        }),
        _ => {}
    }
    let left = pos.step(Dir::West);
    match grid[left] {
        b'-' | b'L' | b'F' => neighbors.push(CoordDir {
            coord: left,
            dir: Dir::West,
        }),
        _ => {}
    }
    if let &[a, b] = &neighbors[..] {
        Some((a, b))
    } else {
        None
    }
}

fn get_start_char(mut a: Dir, mut b: Dir) -> Option<u8> {
    if a as isize > b as isize {
        (a, b) = (b, a)
    }
    match (a, b) {
        (Dir::North, Dir::East) => Some(b'L'),
        (Dir::North, Dir::South) => Some(b'|'),
        (Dir::North, Dir::West) => Some(b'J'),
        (Dir::East, Dir::South) => Some(b'F'),
        (Dir::East, Dir::West) => Some(b'-'),
        (Dir::South, Dir::West) => Some(b'7'),
        _ => None,
    }
}
